### PR TITLE
Replace usage of hadoop.fs.Path with String in Delta Lake

### PR DESCRIPTION
**Pull request by pajaks**
## Description
Origin of this change is to migrate to TrinoFileSystem
Relates to https://github.com/trinodb/trino/issues/16020

Some hadoop.fs.Path will be migrated in following PR as it needs some additional path resolution functions.

## Release notes

(x) This is not user-visible or docs only and no release notes are required.

### CODE REVIEW

**Comment by pajaks** [comment]
@ebyhr Could you run tests with secretes?

**Comment by ebyhr** [comment]
/test-with-secrets sha=fe0df4c21f640800bb37143aba3cc6e24f43f686

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 1]
```diff
@@ -724,13 +724,13 @@ public void createTable(ConnectorSession session, ConnectorTableMetadata tableMe
         }
         Path targetPath = new Path(location);
         ensurePathExists(session, targetPath);
-        Path deltaLogDirectory = getTransactionLogDir(targetPath);
+        String deltaLogDirectory = getTransactionLogDir(location);
```
Comment (alexjo2144) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:727]: Can we migrate `targetPath` two lines up as well?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/StringPathUtil.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+public final class StringPathUtil
+{
+    public static final Pattern DOUBLE_SLASH_TO_REPLEACE = Pattern.compile("(?<!\\w+:/?)//+");
+
+    private StringPathUtil() {}
+
+    public static String stringPathOf(String... pathParts)
+    {
+        String path = String.join(File.separator, pathParts);
```
Comment (alexjo2144) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/StringPathUtil.java:27]: `File.separator` is operating system dependent, but I don't think we want that. Either use / or define a PATH_SEPARATOR constant on this class.
```suggestion
        String path = String.join("/", pathParts);
```

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/writer/TransactionLogWriter.java [hunk 3]
```diff
@@ -107,7 +107,7 @@ public void flush()
         }
 
         String clusterId = commitInfoEntry.get().getCommitInfo().getClusterId();
-        logSynchronizer.write(session, clusterId, logEntry, bos.toByteArray());
+        logSynchronizer.write(session, clusterId, new Path(logEntry), bos.toByteArray());
```
Comment (alexjo2144) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/writer/TransactionLogWriter.java:110]: Can we migrate logSynchronizer as well? can be follow up, I know there are a lot of these

**Review by alexjo2144 — COMMENTED** [review]
Looks good overall, couple small things. 

I'd also dd `Delta` or `Delta Lake` to the commit message somewhere.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/writer/TransactionLogWriter.java (hunk 3 above)
Comment (pajaks) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/writer/TransactionLogWriter.java:110]: I was planning to do that in separate commit as some implementation of `LogSynchronizer` are strongly coupled with `org.apache.hadoop.fs.Path `

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java (hunk 1 above)
Comment (pajaks) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:727]: It's still used in code which will be removed in #16200. So I would wait for that PR to land and remove it during rebase.

**Comment by pajaks** [comment]
rebased to resolve conflicts

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/StringPathUtil.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+import java.util.regex.Pattern;
+
+public final class StringPathUtil
+{
+    public static final String PATH_SEPARATOR = "/";
+    public static final Pattern DOUBLE_SLASH_TO_REPLACE = Pattern.compile("(?<!\\w+:/?)//+");
+
+    private StringPathUtil() {}
+
+    public static String stringPathOf(String... pathParts)
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/StringPathUtil.java:25]: Unit tests seem to be missing for this class.

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/transactionlog/checkpoint/TestTransactionLogTail.java [hunk 5]
```diff
@@ -66,8 +63,7 @@ private List<DeltaLakeTransactionLogEntry> readJsonTransactionLogTails(String ta
             throws Exception
     {
         TrinoFileSystem fileSystem = new HdfsFileSystemFactory(HDFS_ENVIRONMENT).create(SESSION);
-        URI resource = getClass().getClassLoader().getResource(tableLocation).toURI();
-        Path tablePath = new Path(resource);
+        String tablePath = getClass().getClassLoader().getResource(tableLocation).toURI().toString();
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/transactionlog/checkpoint/TestTransactionLogTail.java:66]: Based on the latest changes from ebe2554a the term `path` has been changed to `location`.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/StringPathUtil.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+ * limitations under the License.
+ */
+package io.trino.plugin.deltalake;
+
+import java.util.regex.Pattern;
+
+public final class StringPathUtil
+{
+    public static final String PATH_SEPARATOR = "/";
+    public static final Pattern DOUBLE_SLASH_TO_REPLACE = Pattern.compile("(?<!\\w+:/?)//+");
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/StringPathUtil.java:21]: I need a few unit tests to see whether there is no cheaper alternative to this negative look behind regex.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeWriter.java [hunk 7]
```diff
@@ -198,7 +198,7 @@ private static DeltaLakeJsonFileStatistics readStatistics(
             typeForColumn.put(dataColumnNames.get(i), dataColumnTypes.get(i));
         }
 
-        TrinoInputFile inputFile = fileSystem.newInputFile(new Path(tableLocation, relativeFilePath).toString());
+        TrinoInputFile inputFile = fileSystem.newInputFile(stringPathOf(tableLocation, relativeFilePath));
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeWriter.java:201]: I'm asking naively, why is not enough something like this:

```
new Path(tableLocation, relativeFilePath).toString
```

?

Comment (pajaks) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeWriter.java:201]: > I'm asking naively, why is not enough something like this:
> 
> ```
> new Path(tableLocation, relativeFilePath).toString
> ```
> 
> ?

The idea behind this PR is to remove dependency to hadoop and migrate to TrinoFileSystem as discribed in https://github.com/trinodb/trino/issues/15921

Comment (findinpath) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeWriter.java:201]: Sorry for the confusion. I was thinking that `java.nio.file.Path` was being used in `trino-iceberg` , but  I see that it still uses hadoop `Path`.

**Comment by pajaks** [comment]
Rebased to use 
https://github.com/trinodb/trino/blob/39f030b40f1b5b814de08788f5a6ee95985cc1aa/lib/trino-filesystem/src/main/java/io/trino/filesystem/Locations.java#L22

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 8]
```diff
@@ -825,7 +824,7 @@ public DeltaLakeOutputTableHandle beginCreateTable(ConnectorSession session, Con
         Path targetPath = new Path(location);
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:824]: Did you omit this occurence of `org.apache.hadoop.fs.Path` ?

**Comment by pajaks** [comment]
@ebyhr @findepi Could you run with secrets?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java (hunk 8 above)
Comment (pajaks) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:824]: Those changes are in follow up PR: https://github.com/trinodb/trino/pull/16469

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/transactionlog/checkpoint/TestTransactionLogTail.java [hunk 9]
```diff
@@ -54,10 +52,9 @@ private List<DeltaLakeTransactionLogEntry> updateJsonTransactionLogTails(String
             throws Exception
     {
         TrinoFileSystem fileSystem = new HdfsFileSystemFactory(HDFS_ENVIRONMENT).create(SESSION);
-        URI resource = getClass().getClassLoader().getResource(tableLocation).toURI();
-        Path tablePath = new Path(resource);
-        TransactionLogTail transactionLogTail = TransactionLogTail.loadNewTail(fileSystem, tablePath, Optional.of(10L), Optional.of(12L));
-        Optional<TransactionLogTail> updatedLogTail = transactionLogTail.getUpdatedTail(fileSystem, tablePath);
+        String tableLocationFromResources = getClass().getClassLoader().getResource(tableLocation).toURI().toString();
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/transactionlog/checkpoint/TestTransactionLogTail.java:55]: `tableLocationFromResources` -> `tableLocation`

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/transactionlog/checkpoint/TestTransactionLogTail.java [hunk 10]
```diff
@@ -66,9 +63,8 @@ private List<DeltaLakeTransactionLogEntry> readJsonTransactionLogTails(String ta
             throws Exception
     {
         TrinoFileSystem fileSystem = new HdfsFileSystemFactory(HDFS_ENVIRONMENT).create(SESSION);
-        URI resource = getClass().getClassLoader().getResource(tableLocation).toURI();
-        Path tablePath = new Path(resource);
-        TransactionLogTail transactionLogTail = TransactionLogTail.loadNewTail(fileSystem, tablePath, Optional.of(10L));
+        String tableLocationFromResources = getClass().getClassLoader().getResource(tableLocation).toURI().toString();
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/transactionlog/checkpoint/TestTransactionLogTail.java:66]: same as above

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/transactionlog/checkpoint/TestTransactionLogTail.java (hunk 9 above)
Comment (pajaks) [reply plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/transactionlog/checkpoint/TestTransactionLogTail.java:55]: tableLocation would collide with parameter name

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/transactionlog/checkpoint/TestTransactionLogTail.java (hunk 10 above)
Comment (pajaks) [reply plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/transactionlog/checkpoint/TestTransactionLogTail.java:66]: tableLocation would collide with parameter name

**Review by findinpath — APPROVED** [review]

**Comment by ebyhr** [comment]
/test-with-secrets sha=94a4e3e6bcb2eb8c5f031290930d0c75a66bcac1

**Comment by pajaks** [comment]
CI hit: https://github.com/trinodb/trino/issues/13199 and also  download github action timeout 
@ebyhr Could you rerun?

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/transactionlog/checkpoint/TestTransactionLogTail.java [hunk 11]
```diff
@@ -66,9 +63,8 @@ private List<DeltaLakeTransactionLogEntry> readJsonTransactionLogTails(String ta
...
+        TransactionLogTail transactionLogTail = TransactionLogTail.loadNewTail(fileSystem, tableLocationFromResources, Optional.of(10L));
```
Comment (ebyhr) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/transactionlog/checkpoint/TestTransactionLogTail.java:67]: nit: We can build the resource's location within `testTail` and then simplify `updateJsonTransactionLogTails` & `readJsonTransactionLogTails` methods.
```java
String tableLocation = getClass().getClassLoader().getResource(format("%s/person", dataSource)).toURI().toString();
```

**Review by ebyhr — APPROVED** [review]

**Comment by ebyhr** [comment]
/test-with-secrets sha=94a4e3e6bcb2eb8c5f031290930d0c75a66bcac1

**Comment by pajaks** [comment]
Rebased to include fix for skipped tests in CI

**Pull request merged**